types/wlr_pointer_constraints_v1.c

```diff
@@ -0,0 +1,387 @@
+		struct wl_resource *resource, struct wl_resource *region) {
+	struct wlr_pointer_constraint_v1 *constraint =
+		pointer_constraint_from_resource(resource);
+
+	constraint->pending.region = region;
+}
+
+// only valid for locked_pointer
+static void pointer_constraint_set_cursor_position_hint(struct wl_client *client,
+		struct wl_resource *resource, wl_fixed_t x, wl_fixed_t y) {
+	struct wlr_pointer_constraint_v1 *constraint =
+		pointer_constraint_from_resource(resource);
+
+	constraint->pending.cursor_hint.x = wl_fixed_to_double(x);
+	constraint->pending.cursor_hint.y = wl_fixed_to_double(y);
+	constraint->pending.cursor_hint.valid = true;
+}
+
+static void handle_surface_commit(struct wl_listener *listener, void *data) {
+	struct wlr_pointer_constraint_v1 *constraint =
+		wl_container_of(listener, constraint, surface_commit);
+
+	if (constraint->current.region != constraint->pending.region) {
+		send_unconstrained(constraint);
+	}
+	constraint->current = constraint->pending;
+}
+
+static void handle_surface_destroy(struct wl_listener *listener, void *data) {
+	struct wlr_pointer_constraint_v1 *constraint =
+		wl_container_of(listener, constraint, surface_destroy);
+
+	send_unconstrained(constraint);
+	wl_resource_destroy(constraint->resource); // is this needed?
+}
+
+static const struct zwp_confined_pointer_v1_interface confined_pointer_impl = {
+	.destroy    = resource_destroy,
+	.set_region = pointer_constraint_set_region,
+};
+
+static const struct zwp_locked_pointer_v1_interface locked_pointer_impl = {
+	.destroy                  = resource_destroy,
+	.set_cursor_position_hint = pointer_constraint_set_cursor_position_hint,
+	.set_region               = pointer_constraint_set_region,
+};
+
+static void pointer_constraint_create(struct wl_client *client,
+		struct wl_resource *pointer_constraints_resource, uint32_t id,
+		struct wl_resource *surface_resource, struct wl_resource *pointer,
+		struct wl_resource *region, uint32_t lifetime,
+		enum wlr_pointer_constraint_v1_type type) {
+	struct wlr_pointer_constraints_v1 *pointer_constraints = pointer_constraints_from_resource(pointer_constraints_resource);
+	struct wlr_surface *surface = wlr_surface_from_resource(surface_resource);
+
+	struct wl_list *list = &surface->pointer_constraints;
+
+	int version = wl_resource_get_version(pointer_constraints_resource);
+
+	bool locked_pointer = type == WLR_POINTER_CONSTRAINT_V1_LOCKED;
+
+	struct wl_resource *resource = locked_pointer ?
+		wl_resource_create(client, &zwp_locked_pointer_v1_interface, version, id) :
+		wl_resource_create(client, &zwp_confined_pointer_v1_interface, version, id);
+
+	if (resource == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	struct wlr_seat_client *seat_client = wlr_seat_client_from_pointer_resource(pointer);
+
+	struct wlr_pointer_constraint_v1 *constraint;
+	wl_list_for_each(constraint, list, link) {
+		if (constraint->seat_client == seat_client) {
```

Comment:
This won't catch the case when a different seat client has a constraint on the pointer. A seat client is a client connection to a seat and there can be more than one per seat. Was that the intent here?

Reply:
It won't catch a case where two separate seat clients of the same seat hold a pointer constraint on the same surface, although as far as I've understood, a surface is specific to a client, so it's not possible for this to occur, since a seat client of a client that does not own the surface will not be able to request a pointer constraint on the surface, unless there is some way to share surfaces which I am not aware of?